### ISSUE TITLE
DLC.link: Bitcoin-native, non-custodial transactions

### ISSUE THREAD

**Issue by Akibalogh** [stacksgov/Stacks-Grant-Launchpad#289]
## **Background**

*What problems do you aim to solve? How does it serve the mission of a user owned internet?*

Stacks does not have a reliable service to provide its Layer-1 blockchain secure and accurate off-chain pricing data. For a robust, secure, and trust-less decentralized finance (DeFi) experience on applications such as [ALEX](https://www.alexgo.io/) and [Arkadiko](https://arkadiko.finance/), which operate on the Stacks blockchain, the need for a reliable on-chain price oracle is a critical component.

The interest and desire for a non-custodial BTC borrowing/lending protocol is growing. The use of a stable oracle solution could be used to provide accurate price data which then allows the proper margin requirements, loan interest payments, or liquidations to execute. 

Other tasks such as routing payments/transactions on the Stacks blockchain, require the interaction of a third-party. To keep with the mission and goal of BTC and wealth self-custody, a partnership between the Stacks Foundation and DLC.link can explore ways to utilize technology to allow for this to happen. 

## **Project Overview**

*What solution are you providing? Who will it serve?*

The use of Discreet Log Contracts (DLCs) in combination with a trusted oracle solution such as Chainlink, solves the problem of unlocking bitcoin liquidity without involving a custodian or having to trust a third-party. This technology will serve anyone who is willing to only use bitcoin as collateral to open, pay interest, and close a loan agreement in a way that aligns with the bitcoin philosophy around a trustless monetary system that settles things in a programmatically, immutable way.

![DLC Architecture Diagram](https://lh6.googleusercontent.com/TFnXFRhTeE_4vxcQIeCLk6RehTJV0jWsTWyBb7iD8gP_8WXl9zcY64WGMBbHPYDLWFLea41mDJctg6199gVy08UA88K9XIgRRgnb13nMAfXl4qhpwRGzgZ1MQfz68o8Fu1dPHyc)

## **Scope**

*What are the components or technical specs of the project? What will the final deliverable look like? How will you measure success?*

Summary of Scope Items:
1. Create a Chainlink oracle contract on the Stacks blockchain 
2. Implement a DLC on Stacks utilizing a price-feed oracle 
3. Gather market information around other off-chain data that Stacks could utilize
4. Research enhancements to the Clarity smart contract language that would benefit development on Stacks
5. Create documentation and training materials

**Item 1: Create a Chainlink oracle smart contract on the Stacks blockchain**
Create a Chainlink [BTC-USD](https://data.chain.link/ethereum/mainnet/crypto-usd/btc-usd) price feed contract that lives on the Stacks blockchain. To do this requires developing a platform to interact with Chainlink external adaptors and the actual adaptor itself. In addition, this scope item includes research and information gathering around how existing smart contracts on Stacks would utilize price information.

**Item 2: Implement a DLC on Stacks utilizing a price-feed oracle**
Create a smart contract that is deployed when a DLC is created with the following properties:
* Contains the UUID of the associated DLCLink event
* Contains the address of the Chainlink data feed in which the DLC outcome requires
* Contains the strike/execution price of the reference asset within the DLC
* The contract is connected to the address of the initiator (borrower) allowing the initiator to close the contract at any time, e.g. returning borrowed USDA and returning the BTC collateral at the current price of BTC
* The contract is connected to the counterparty (lender) so that counterparty can close the contract at any time, e.g freeing the borrower from any future liabilities and returning some/all of the BTC collateral, depending on the initial agreement
* The contract will incorporate a mechanism to periodically check the price of the underlying asset utilizing a Chainlink oracle to decide whether certain logic in the smart contract is executed.

**Item 3: Research applicability for DLCs to solve pain points for Stacks teams**
Initiate conversations with active Stacks users and protocols to gather information around other potential data that would be helpful to bring into the Stacks ecosystem. Chainlink currently has weather data, verifiable randomness, and other external adapters which allow for non-price information to be vetted and printed using a decentralized oracle network.

**Item 4: Research enhancements to the Clarity smart contract language that would benefit development on Stacks**
Certain packages, plug-ins, and contracts will most likely need to be created for a successful integration with the Chainlink oracle network and DLC connectivity. The team is interested in tracking what future needs/enhancements can be made to achieve easier widespread use of the DLC.link bridge and oracle data. Call back functionality for the creation of DLCs on Stacks may be a significant enhancement to open more use cases. 

**Item 5: Create documentation and training materials**
For the above items, technical documentation will be created to allow essential information around the smart-contacts, services, code, etc. to be available for those who need to utilize it. Training materials as well as “Getting Started” guides will be built for developers and users to become familiar with DLC.link and understand the capabilities it provides.    

## **Budget and Milestones**
*What grant amount are you seeking? How long will the project take in hours? If more than 20, please break down the project into milestones, with a clear output (e.g., low-fi mockup, MVP with two features) and include the estimated work hours for each milestone.*

Total Grant Request: $250,000

Scope of work, deliverables, and grant amount of total:
1. Facilitate the creation of a Chainlink oracle on the Stacks blockchain using the generalizable OCR model
    1. Milestones:
        1. Assess work done to date on existing CL-Stacks grant to Tintash
        2. Build the business case for CL to prioritize the work internally
        3. Building the DLC Chainlink adapter
        4. Constructing the initial technology platform using Chainlink External Adapters
        5. Launching an open-source prototype that demonstrates proof of concept and can be used by end-users to gather feedback and future requirements
    2. Total grant requested: 600 hours / $75k
   
2. Implement a DLC on Stacks
    1. Milestones:
        1. Adapt the DLC work to use the CL oracle from Item 1
        2. Develop and launch DLC.link oracle contract
        3. Integrate CL data feeds for generalized DLC contracts.
        4. Enable DLC.link bridge to operate on Stacks
        5. Initial security sweep/audit of smart contract logic
        6. Experiment with Stacks wallet integrations
        7. Evangelize to bring in initial users and validate that the solution works for Stacks dApps
    2. Total grant requested: 600 hours / $75k
  
3. Research applicability for DLCs to solve pain points for Stacks teams
    1. Milestones:
        1. Interview Stacks teams to validate types of events data required (Zest)
        2. Determine viability of DLCs to support swap liquidity pools without reliance on 3rd party centralized exchange platforms such as Binance
        3. Gather information around the utility of currency hedging and derivatives powered by DLCs for treasury management (StackerDAO) and others in the Stacks ecosystem
    2. Total grant requested: 200 hours / $25k

4. Research regarding adding Clarity language enhancements to support launching DLCs
    1. Milestones:
        1. DLCLink engineers collaborate with Stacks/Clarity engineers to discuss feasibility and if possible design a library to open DLC contracts from Stacks/Clarity
        2. Do the same as above for triggering a callback function in Clarity when a DLC is funded
    2. Total grant requested: 200 hours / $25k

5. Create documentation and training materials
    1. Milestones:
        1. Create complete process flow diagrams and code visualizations for the Chainlink oracle integration on Stacks
        2. Create a document outlining the process of how the DLC operates on the Stacks blockchain
        3. Create a guide on how to utilize the DLC.link oracle for price information in Clarity smart contracts 
        4. Provide commentary around discussions regarding the further use cases for price feed and other oracle services on Stacks 
    2. Total grant requested: 400 hours / $50k

## **Team**
*Who is building this? What relevant experience do you bring to this project? Are there skills sets you are missing that you are seeking from the community? Please share links to previous work.*

Aki Balogh (https://www.linkedin.com/in/akibalogh/)

Aki invented the predominant method for content optimization for SEO via topic modeling and founded MarketMuse to bring it to market. MarketMuse is an AI-driven content intelligence platform which was named the “undisputed leader in content optimization” in 2019. Previously, Aki focused on scalable data infrastructures at InfiniDB and OpenView Venture Partners and designed decision-support systems for Fortune 500 clients at Diamond.

 

Jesse Eisenberg (https://www.linkedin.com/in/jesses16)

Jesse has served in a number of software engineering and engineering management roles, spanning Pivotal Labs, MoPub and Twitter. He’s a serial entrepreneur excited to put his knowledge of blockchain to work.

Dan von Kohorn (https://www.linkedin.com/in/vonkohorn/)

Dan spans several roles: head of Enterprise Solutions at Chainlink, serial entrepreneur (16 companies and counting), Advisor to ConsenSys and Managing Partner at Broom Ventures.  Currently part-time CEO at IPIQ, an NFT platform to secure intellectual property. Deep background in quantitative finance.

Laszlo Mari (https://www.linkedin.com/in/laszlolm/)

Laszlo runs Dakai.io, a development shop based in Hungary that is a preferred development partner for Binance, the government of El Salvador, AlphaPoint and others.

Matt Bombard (https://www.linkedin.com/in/matthewbombard/)

Matt comes from a financial services background in technology audit. He is passionate about blockchain technology, emerging DeFi challenges, financial derivatives, and analytics.

**Advisors:**

Bernat Nacsa (https://www.linkedin.com/in/bernatnacsa/)

Bernat is a VC with Day One Capital, but he also is interested in collaborating with startups to build out their early visions. He brings his broad understanding of DeFi as well as academic knowledge in financial instruments to the team.

Pablo Fernandez (https://www.linkedin.com/in/pablofer)

Pablo is Director of R&D at Swan Finance. His prior experience spans Bitcoin Beach, Drift, HotelTonight and a number of startups.

## **Risks**

*What dependencies or obstacles do you anticipate? What contingency plans do you have in place?*

Since Clarity is a relatively new language, our knowledge around the Clarity smart-contract programming language may be limited.  That said, we’ve been told that dev shops are able to on ramp to in within a 4-week period.

Because Clarity and STX only have read-only access to the Bitcoin network, there may need to be a permissions workaround (e.g. using wrapped BTC) to settle transactions. This may present unexpected obstacles, but is an issue that Trust Machines is actively working on.

Chainlink Labs work is required for testing and integration to develop a CL oracle on Stacks in production. Miscommunications or delays at Chainlink Labs will directly affect our ability to deliver Chainlink oracle value.

We’re assuming that the generalizable OCR model at Chainlink has been widely adopted and can be utilized to build upon. The current implementation of OCR only supports EVM-based systems. Delays in its completion would become a bottleneck for us as well.

## **Community and Supporting Materials**

*Do you have previous projects, code commits, or experiences that are relevant to this application? What community feedback or input have you received? How do you plan to share your plan to the community over time and as the final deliverable?*

Every two weeks, we plan to hold community calls so as to update on progress against milestones.  We’ll also actively participate on the Stacks Discord and will be available to answer questions.  Within a week, the #dlc-chainlink channel on Discord has already gathered 24 members.  (Link: [https://discord.com/channels/621759717756370964/940317219177525358](https://discord.com/channels/621759717756370964/940317219177525358))

The DLC.link team previously received a grant from Chainlink to bridge Chainlink oracle data to Discreet Log Contracts on the Bitcoin network.  That work is largely aligned with this initiative. For the announcement, see: [https://blog.chain.link/dlc-link-chainlink-grant-bitcoin-discreet-log-contracts/](https://blog.chain.link/dlc-link-chainlink-grant-bitcoin-discreet-log-contracts/)

**Selected Dakai Case studies:**

* **Binance (2021):** Built an NFT marketplace ([Featured.market](http://featured.market/)) that featured launches from Formula-E, Burberry, Balmain and other big brands. We built it in 2 weeks as the previous team left the project
* **Chivo (2021):** A mobile application for the El Salvador government that 6 million people use day-to-day to save and transfer USD and Bitcoin. It was a failing project as the previous team left severe issues in for launch. We fixed the application in 9 weeks, launched a new version, and it scaled to 6 million users within 2-3 weeks without any critical issues in this period.
* **MDAX/Peak6 (2021):** We built an NFT platform and Private Ethereum environment with many regulatory safeguards so that the partners of Peak6 (Disney, Pixar, Universal Studios, Marvel, Warner Bros) can launch movie pieces and exclusive content in a regulated environment.
* **GeneUS (2021):** Dakai built a genetical sequencing platform to diagnose rare diseases out of raw DNA data with a social component around it. The platform is still awaiting regulator approval to launch
* **PortfoLion (2021):** Dakai advises PortfoLion (one of the largest VC firms Central Europe ($500M AUM)) & evaluates their potential investments from the tech point of view.
* **Glimpse (2020):** We helped Glimpse create an innovative set of products around scraping the internet (Tiktok, Reddit, Google, Twitter among others) to predict trends sometimes half a year in advance. The Enterprise version of the application is used by Google among others to spot trends on their own platform. We also helped prepare private consulting materials for A16Z, Sequoia and advised these firms on several investments they had. (One public material we released together with A16Z is: [https://a16z.com/2020/09/16/fintech-covid-effect/](https://a16z.com/2020/09/16/fintech-covid-effect/))

**Comment by owenstrevor** [comment]
Great to see this proposal.

Following-up from the grant meeting this morning, it would be great to see a more agile approach to deployment of this roadmap. Instead of working on everything in parallel with a waterfall-style release at the end, it would be great to know in what order different aspects of this project will be available so that our startups can utilize them!

**Comment by tycho1212** [comment]
Hi guys! For one, I'm very excited that a team is going to be pushing the Chainlink integration to completion. Even if this grant achieves just that, it's a win.

Secondly, from the Zest Protocol perspective we'd love to see something in Item 2 on operating a DLC using inputs from Clarity. To use DLCs for managed BTC liquidity pools, we'll need to give the pool manager (pool delegate) the ability to unlock escrowed BTC to a borrower. This might be consensus-breaking - if it is, we'd want to slot it into Stacks 2.1 which will likely be ready by Q3. I'd be happy to provide some more details on how Zest Protocol intends to use DLCs on a call.

**Comment by castig** [comment]
Excellent proposal! And a solid team behind it. I'd be great if you could keep us on to date on your progress either here, or through an email newsletter. Looking forward to hearing more.

**Comment by FriendsFerdinand** [comment]
I understand that the current proposal only implements an oracle with a price feed (such as BTC/USD). I think it would be beneficial to generalize the type of oracle so that we can implement an oracle for events that happen on the Stacks blockchain.

**Comment by csgui** [comment]
Isn't publicly available?  https://discord.com/channels/621759717756370964/940317219177525358

**Comment by Akibalogh** [comment]
> Isn't publicly available? https://discord.com/channels/621759717756370964/940317219177525358

It's not an open channel currently, so ping @will-at-stacks for an invite.  But at this point it could probably be moved under Developers. There is a chainlink-testing channel that could be merged into this.

**Comment by Akibalogh** [comment]
> I understand that the current proposal only implements an oracle with a price feed (such as BTC/USD). I think it would be beneficial to generalize the type of oracle so that we can implement an oracle for events that happen on the Stacks blockchain.

Thanks for the feedback.  We intended to capture events as "non-financial data" in milestone #3.  I'll add a subtask to include that explicitly.

**Comment by Akibalogh** [comment]
> Excellent proposal! And a solid team behind it. I'd be great if you could keep us on to date on your progress either here, or through an email newsletter. Looking forward to hearing more.

Thanks!  Please sign up at https://dlc.link to receive project-specific content (included in "documentation and training" in milestone 5)

**Comment by Akibalogh** [comment]
> Hi guys! For one, I'm very excited that a team is going to be pushing the Chainlink integration to completion. Even if this grant achieves just that, it's a win.
> 
> Secondly, from the Zest Protocol perspective we'd love to see something in Item 2 on operating a DLC using inputs from Clarity. To use DLCs for managed BTC liquidity pools, we'll need to give the pool manager (pool delegate) the ability to unlock escrowed BTC to a borrower. This might be consensus-breaking - if it is, we'd want to slot it into Stacks 2.1 which will likely be ready by Q3. I'd be happy to provide some more details on how Zest Protocol intends to use DLCs on a call.

Thank you for the support!  This is intended to be covered in items 2 and 4, though our current wording is imprecise.  We have a call set this Thurs, so let's dig in then.  We'd also love your advice on what's coming up in Stacks 2.1.

**Comment by Akibalogh** [comment]
Thank you for the detailed comments and apologies for my delay in responding. I wanted to get back to you when I had meaningful information to share. Feel free to comment further in GitHub and we can also set time live as well (I'm at akibalogh#1043).

> Following-up from the grant meeting this morning, it would be great to see a more agile approach to deployment of this roadmap. Instead of working on everything in parallel with a waterfall-style release at the end, it would be great to know in what order different aspects of this project will be available so that our startups can utilize them!

We certainly didn't want to create the impression that this would be delivered as a waterfall! Our understanding of the goal of last Friday's presentation was to begin to formally start working on it and bill hours against the work. Now that we're engaging technical resources, we're able to provide a technical roadmap that has been validated by Dakai and our team. This Thursday, we're provide Stacks Foundation an initial roadmap with upcoming due dates so that they can approve the partial progress made so far. Some of the more distant deliverables will be marked TBD. As a sneak preview, the first delivery date we're aiming for is a prototype CL-Stacks bridge by Thurs March 31, so that we can also present it at the Stacks event in Miami the following week.

> For a grant of this size, I would also recommend more milestones than just 5.

We feel comfortable with the 5 overarching goals and the known subtasks. Since this project has heavy product development components, the number and complexity of subtasks will increase as we make progress in our research and product validation work.

> Especially for the first and second milestones, I would expect to see something tangible delivered for less than $75k to see that this project is being executed on. 

Given the scope of what we're looking to do, I don't see an issue. Stacks teams are eagerly anticipating a Chainlink oracle and see valuable use cases for DLCs as well. Last year, the challenged launch of the CL-Stacks adapter and the issue that Arkadiko ran into both clearly show that this problem is a major barrier to Stacks' further development. Given the business value unlocked by this project, we see a lot of upside. The resources for this engagement are scoped in an appropriate manner.

> For the team section, I'm a little confused who's actually working on this full-time? It appears several team members have other obligations but I'm not sure. Is DLC.link a full-time company you are all dedicated to? Several members have other companies on their linkedin pages.

To keep costs low, none of us are currently full-time on DLC.link. We all have the ability to ramp up to full-time, but we wanted to present a more affordable package by diversifying our time over multiple projects. Taking just me as an example: as a combined project leader/product manager/project manager, I'm able to provide 40 billable hours per week, but I am aiming to only work 20-30 billable hours so as to keep costs within the indicated budget, while still hitting the milestsones indicated. Though we don't have a track record at Stacks yet, our professional backgrounds indicate that we're experienced resources who have a history of following-through on our commitments.

> Milestones 3-5 seem a bit expensive without more information.
>   --Milestone 3 - $25k and 200 hours for doing research.
>   --Milestone 4 - $25k and 200 hours to discuss something and see if it's feasible.
>   --Milestone 5 - $50k and 400 for documentation and training materials.

This critique of Milestone 3 is completely fair, since we've only provided limited information to date. In general, these milestones are getting filled out as we work with teams at Stacks and are going to get updated as we progress. Case in point, Milestone 3 was just updated today with new requests from two Stacks teams.

Our request for Milestone 4 is predicated on the fact that there are unknown unknowns as it pertains to Clarity language extensions. I expect this milestone to evolve as we get through Milestones 1 & 2. Essentially, this research enables us to build something that Stacks teams and portfolio companies can implement in practice. As our collective understanding of the development need evolves, we'll evolve this milestone to reflect that.

Milestone 5 helps address a specific request from the Stacks Foundation. As technical founders, it's easy to deprioritize documentation and training. More broadly, onboarding new users is a huge problem at crypto in general. We also anticipate that there will be marketing activities (such as the newsletter that was asked for in a previous question) that is very necessary but can easily get skipped if not resourced adequately.

I hope that this provides more detail on our thought process. Again, happy to jump on live and talk through this as well.

**Comment by will-at-stacks** [comment]
@Akibalogh -

Thanks for all your work on this.  As per my email, we will discuss these revisions internally tomorrow.  Adding this note for the community record:
-----
If approved, would the following payment and progress review schedule be amenable to you?

- Initial payment + retainer, $50k at approval
- (4) $50k payments paid at the end of each month, provided sufficient progress is made as determined at the monthly progress review.

**Comment by Akibalogh** [comment]
@owenstrevor and All - please find our updated roadmap here:
https://docs.google.com/document/u/3/d/1tiUIT_7BXgrRJA6E1vneCMYvQeGzpxSzEb0Dm9MlXGo/edit

>If approved, would the following payment and progress review schedule be amenable to you? Initial payment + retainer, $50k at approval. (4) $50k payments paid at the end of each month, provided sufficient progress is made as determined at the monthly progress review.

@will-at-stacks thank you - that works great for us

**Comment by will-at-stacks** [comment]
@Akibalogh 

Pleased to let you know that your grant is approved!

- We will work to get the contract in place and disburse the 1st of the (5) $50k progress payments in the next few days. 
- Per our conversation Trevor, Tycho, and myself will serve as a review committee for the duration of your grant.  
- All progress payments will be released upon satisfactory completion of 'acceptance criteria' established by your team and approved by myself @owenstrevor and @tycho1212.   
- Please provide acceptance criteria for the first months work at your earliest convenience. 
- I will connect you with Shakti to help answer any payment related questions.

LFG!!!

**Comment by will-at-stacks** [comment]
@Akibalogh 
Great meeting with your team just now to discuss your Milestone (month 1) progress.  Per our call, it sounds like due to start-up mobilization and a longer than anticipated discovery phase items 3-5 are going to slide from Milestone 1 deliverables to later in the project.  That being the case we are still agreeing to release the next $50k monthly (month2) payment so you can move to make up some ground.  However, as discussed, we are going to prioritize having a working DLC use case as a prerequisite for releasing the month 3 $50k.  If your team can also continue to make progress on the Clarity improvements and Chainlink fact finding that would be great.

Best,
Will

Facilitate the creation of a Chainlink oracle on the Stacks blockchain using the generalizable OCR model:
1. Assess work done to date on existing CL-Stacks grant to Tintash
2. Build the business case for CL to prioritize the work internally
3. Building the DLC Chainlink adapter
4. Constructing the initial technology platform using Chainlink External Adapters
5. Launching an open-source prototype that demonstrates proof of concept and can be used by end-users to gather feedback and future requirements

**Comment by will-at-stacks** [comment]
Hello and thank you for participating in the Stacks Foundation Grants Program!

We are in the process of migrating from GitHub to the new [Grants Dashboard](https://grants.stacks.org/). In order to complete your grant, you will need to submit any remaining Progress Review and/or Final Review requests through the Dashboard in order to receive your remaining payments.

- If you have questions about the Grants Dashboard in general please email [ambassadors@stacks.org](mailto:ambassadors@stacks.org).
- If you have questions regarding payments and/or contracts please email [payments@stacks.org](mailto:payments@stacks.org).

Lastly, please note we are marking this grant 'closed' on GitHub for organizational purposes, but it is still 'open' on the Grants Dashboard.

Thanks and we hope to continue to support your efforts with additional grants!

Best,
Will

**Comment by Hero-Gamer** [comment]
How can I follow the milestones of this Grant?

**Comment by sosaucily** [comment]
Hey @Hero-Gamer good question, and thanks for taking interest and following up!

We don't have a public roadmap that clearly connects to these milestones, but I can at least give some updates here.

**Milestone 1: Chainlink price-feed oracle on Stacks.** 
This was completed last year, there was an announcement and a public discord channel set up for it. However there wasn't much interest from within the team, and unfortunately we haven't had luck in getting Chainlink to formally support the effort as we had hoped, which means that product lags behind the state-of-the-art at this point unfortunately. In fact we're even leveraging the Redstone solution currently for our DLC.Link contracts now, although we could switch to Chainlink in the future if it gets better support.

**Milestone 2: Implement a DLC on Stacks.**
We have completed the majority of items here, and are working on getting a formal integration with Arkadiko to create a lending product leveraging native BTC and defi lending. We're nearing the end of integration development with Arkadiko, and currently the main blocker for releasing that to the public in a beta format is our DLC integration into the Xverse wallet. All that is estimated to complete sometime in Q1 of '23. 

**Milestone 3: Research applicability for DLCs to solve pain points for Stacks teams**
Completed last year. Last year we reached out to all the teams building Stacks to discuss if DLCs could fit. We're currently working closest with Arkadiko and Zest, but staying in touch with the others too!

**Milestone 4: Research regarding adding Clarity language enhancements to support launching DLCs**
Done. We believe we've done a good job of leveraging Clarity well for teams that want to integrate with DLCs. We have 3 traits publically deployed in Testnet along with our contract, which make creating and closing DLCs fast and easy. But of course this will improve over time as we see how teams are really using DLCs and our code.

**Milestone 5: Create documentation and training materials**
An ongoing effort, but that will be improved and formally released in Q1 as we continue our integrations into the Xverse wallet and Arkadiko system. We have product focused docs currently, and will be adding more coding focused docs on https://docs.dlc.link soon!

Let me know if you have any more questions, or think any of this information should be released publically in discord, or via some marketing channel, etc.
